### PR TITLE
add build sdkconfig support

### LINKED ISSUE #855: [Feature Request]: support sdkconfig.defaults instead of sdkconfig (VSC-1019)
For multiple build configs:
https://github.com/espressif/esp-idf/blob/master/examples/build_system/cmake/multi_config/README.md

In that recommended example, the `sdkconfig` file is placed in the `build` directory, and `sdkconfig.defaults` is used in the parent directory.

It seems VSC does not recognize this configuration and keeps putting a new `sdkconfig` file into the parent directory. 

I think VSC should open the `sdkconfig.defaults` file if it exists, and save to that file only, if it exists.

### PR DESCRIPTION
**Pull request by brianignacio5**
## Description

Add support for multiple build configuration as done in [esp-idf cmake multi_config example](https://github.com/espressif/esp-idf/tree/master/examples/build_system/cmake/multi_config).

Load sdkconfig path as follows:
1) From current workspace folder root directory `CMakeLists.txt` SDKCONFIG variable else
2) From `SDKCONFIG` environment variable (either system or extension defined in `idf.customExtraVars` setting) else
3) From current workspace folder root directory `sdkconfig` file.

Also add `idf.sdkconfigDefaults` to define the set of SDKCONFIG_DEFAULTS to use to initially create the sdkconfig.

Fix #855 

## Type of change

Please delete options that are not relevant.
- New feature (non-breaking change which adds functionality)

## How has this been tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- Manual testing building and running SDK Configuration editor.

**Test Configuration**:
* ESP-IDF Version: 4.4. 5.1
* OS (Windows,Linux and macOS): macOS

## Checklist
- [x] PR Self Reviewed
- [x] Applied Code formatting
- [x] Added Documentation
- [ ] Added Unit Test
- [ ] Verified on all platforms - Windows,Linux and macOS

### CODE REVIEW

**Comment by chipweinberger** [comment]
Thanks for adding this feature Brian! 

> Load sdkconfig path as follows:
>     From current workspace folder root directory CMakeLists.txt SDKCONFIG variable else

I am trying to understand, how does this support "multiple" build configuration? It seems like this PR is more about supporting a custom SDKCONFIG _path_?

Is the idea that I am supposed to change the CMakeLists.txt SDKCONFIG variable myself? i.e. without a UI

**Comment by chipweinberger** [comment]
I just found and read your documentation (thanks)!

To answer my own question, there is no UI for switching multiple build configs at this time. 

Screen Shot 2023-04-07 at 2 09 34 PM

**Comment by brianignacio5** [comment]
Thanks for your feedback @chipweinberger Still need to have more links to docs in the main readme to make it more visible.

There is also a tutorial for the Project Configuration Editor in [here](https://github.com/espressif/vscode-esp-idf-extension/blob/master/docs/tutorial/multiple_config.md) about using this UI you mentioned. If you have any feedback let us know !

**Comment by chipweinberger** [comment]
One thing I still dont understand. 

I've set my sdkconfig vars as such:

Screenshot 2023-04-26 at 4 02 00 PM

But I still don't understand, which file does the SDK Config Editor load and save to?

Screenshot 2023-04-26 at 4 01 42 PM

Is there any way to Edit my `sdkconfig.defaults` or `sdkconfig.prod` file using the UI?

**Comment by chipweinberger** [comment]
IMO, the SDK Configuration Editor should be *per file*.

Instead of UI table title saying "SDK Configuration Editor", it should say "sdkconfig.default" or "sdkconfig.prod" depending on the file you are currently editing.

I also think the `idf.sdkconfigDefaults` variable does not make much sense.

It should support multiple configurations.

i.e. 

```
idf.sdkconfigDefaults : {
     "dev: [
        "/Volumes/User/MBP-Google-Drive/jamcorder/firmware/jamcorder-firmware/jamcorder_app/sdkconfig.defaults"
    ]
     "prod: [
        "/Volumes/User/MBP-Google-Drive/jamcorder/firmware/jamcorder-firmware/jamcorder_app/sdkconfig.defaults"
        "/Volumes/User/MBP-Google-Drive/jamcorder/firmware/jamcorder-firmware/jamcorder_app/sdkconfig.prod"
    ]
}
```

Next to the Gear Icon should be a label (in this case it would say "dev" or "prod" since that is what I've named my 2 configurations above)

In the SDK Configuration Editor, the title should say which file you are currently editing. And of course, we need a way to switch which file we are editing.

**Comment by brianignacio5** [comment]
Hi @chipweinberger so the idea of sdkconfig.defaults is to **build** an sdkconfig file. 

Let's say you have 2 products, **prod1** and **prod2**. Each product has a single resulting sdkconfig file in `${workspaceFolder}/build_prod1/sdkconfig` and `${workspaceFolder}/build_prod2/sdkconfig`. 

1) `${workspaceFolder}/build_prod1/sdkconfig` is built from  `sdkconfig.defaults; [sdkconfig.common, sdkconfig.prod1]`. See how `sdkconfig.common` just list a set of common settings for prod1 and prod2 while sdkconfig.prod1 have product 1 specific settings.

2) `build_prod2/sdkconfig` is built from  `sdkconfig.defaults; [sdkconfig.common, sdkconfig.prod2]`. See how `sdkconfig.common` just list a set of common settings for prod1 and prod2 while sdkconfig.prod2 have product 2 specific settings.

3) For the tutorial example, the location of sdkconfig file is determine in the CMakeLists.txt SDKCONFIG variable.

4) When you choose a profile, say prod1, you are setting the build directory to `build_prod1` so when you try to build or use the SDK Configuration Editor it will use the `build_prod1/sdkconfig` to save settings and build your binaries in `${workspaceFolder}/build_prod1`.

The currently selected profile is shown in the status bar at the bottom of the Visual Studio Code.

Does this help @chipweinberger ? We should clarify a bit more of this in the documentation and tutorials.

**Comment by brianignacio5** [comment]
By the way, We are also adding the sdkconfig file location and IDF_TARGET to such profile configuration. So will try to explain better the difference between `sdkconfig` and `sdkconfig.defaults`

**Comment by chipweinberger** [comment]
## question 1

> Does this help @chipweinberger ? 

Yes, it mostly makes sense to me the **_build process_** you are describing.

But I want to know about the **_editing workflow_**. How do I edit a specific sdkconfig file using the "SDK Config Editor"?

## question 2

You say "sdkconfig.common". How do I set the name of "sdkconfig.common"? Does it have to be called "sdkconfig.common" or can I call it something else?

## question 3

The json format I describe here makes more intuitive sense to me. But how do I convert these settings I'm describing below into the settings you are describing? What should I set?

```
idf.sdkconfigDefaults : {
     "dev: [
        "/Volumes/User/MBP-Google-Drive/jamcorder/firmware/jamcorder-firmware/jamcorder_app/sdkconfig.defaults"
    ]
     "prod: [
        "/Volumes/User/MBP-Google-Drive/jamcorder/firmware/jamcorder-firmware/jamcorder_app/sdkconfig.defaults"
        "/Volumes/User/MBP-Google-Drive/jamcorder/firmware/jamcorder-firmware/jamcorder_app/sdkconfig.prod"
    ]
}

```

## question 4 & 5

Does the json description format above make sense to you? Maybe you can consider if this feature could made simpler to understand?

**Comment by brianignacio5** [comment]
For the current behavior, 1) Select a profile `prod1` which has a `build_prod1` build directory. The sdconfig path will be resolved by replacing CMAKE_BINARY_DIR  in `${CMAKE_BINARY_DIR}/sdkconfig` from [CMakeLists.txt](https://github.com/espressif/esp-idf/blob/master/examples/build_system/cmake/multi_config/CMakeLists.txt#L6) with the current profile build directory path. Any changes you make in the SDK Configuration editor will be saved in `build_prod1/sdconfig`.

> **NOTE:** `Will add the sdkconfig as another setting in the project configuration wizard so the user can manually define the sdkconfig used for development.`

2) `sdkconfig.common` is just the name of the file listed in sdkconfigDefaults array. It can be any name. The way sdkconfig is generated from these defaults is resolving these sdkconfig files in order. So

`sdkconfigDefaults; [sdkconfig.common, sdkconfig.prod1]` will use values from `sdkconfig.common` (which a typical use case will be to put common settings for all profiles here) and then load on top values of `sdkconfig.prod1` (which would have a specific product setting).

The example is just to illustrate why to use sdkconfigDefaults. You would define these values when you want to distribute the source code in a cleaner way without ESP-IDF generated values in resulting sdkconfig. If you look at the [Multiple configuration example](https://github.com/espressif/esp-idf/blob/master/examples/build_system/cmake/multi_config) the sdkconfig.prod1 and sdkconfig.prod2 specify unique values but resulting sdkconfig files include ESP-IDF values.

3) I think it might be better to think that the use of `sdkconfigDefaults` for production only and use sdkconfig for  development purposes. Usually the sdkconfig is generated by the sdk configuration editor and build process. You modify `sdkconfig` and build again. 

JSON would be

```json
"product1": {
   "idf.sdkconfigDefaults" : {
      "${workspaceFolder}/sdkconfig.common"
      "${workspaceFolder}/sdkconfig.prod1"
   },
   "idf.sdkconfig": "${workspaceFolder}/sdkconfig.product1",
   "idf.buildDirectoryPath": "${workspaceFolder}/product1_build"
},
"product2": {
   "idf.sdkconfigDefaults" : {
      "${workspaceFolder}/sdkconfig.common"
      "${workspaceFolder}/sdkconfig.prod2"
   },
   "idf.sdkconfig": "${workspaceFolder}/sdkconfig.product2",
   "idf.buildDirectoryPath": "${workspaceFolder}/product2_build"
}
```

4) I understand the idea you are referring. But the purposes of these files is more for production while sdkconfig is more for development process.

5) This files are used in many CMake projects so in terminology it would be best to keep them close as original intent. We do need to make it easier to understand what are there for better though.

**Comment by chipweinberger** [comment]
I still do not really understand very well. 

## Question 1

> For the current behavior, 1) Select a profile prod1 which has a build_prod1 build directory. The sdkconfig path will be resolved by replacing CMAKE_BINARY_DIR in ${CMAKE_BINARY_DIR}/sdkconfig from [CMakeLists.txt](https://github.com/espressif/esp-idf/blob/master/examples/build_system/cmake/multi_config/CMakeLists.txt#L6) with the current profile build directory path. Any changes you make in the SDK Configuration editor will be saved in build_prod1/sdkconfig.

> "Any changes you make in the SDK Configuration editor will be saved in build_prod1/sdkconfig.", 

That does not make sense to me. The build folder is ephemeral, and gets deleted on full clean. If you "save" to the build folder, then it is not really saved at all. It will be deleted!

It should save to our workspace folder! I want "git status" to show the change, so that I can commit it! 

## question 3

> I think it might be better to think that the use of sdkconfigDefaults for production only and use sdkconfig for development purposes. Usually the sdkconfig is generated by the sdk configuration editor and build process. You modify sdkconfig and build again.

This seems to be the biggest disconnect. I really don't understand what `sdkconfigDefaults` does. I don't understand how it affects building. I dont understand how it affects the "SDK Config Editor". I don't understand how to switch multiple configs.

## summary

Overall, I just don't really _get it._ I don't understand how it is "supposed" to work. The PR made things more complicated, and AFAICT, made things worse for me.

Before `SDK Config Editor` created a `sdkconfig` file in my workspace. Things were simple. I understood it. 

After I switched my product to [multi config](https://github.com/espressif/esp-idf/tree/master/examples/build_system/cmake/multi_config), the VSC extension did not support it at first, so I continued to also use a `sdkconfig` file in my workspace. I would manually copy settings to me `sdkconfig.defaults` and `sdkconfig.prod` as needed. Things were simple. I understood it.

AFAICT, the PR makes things worse. The `sdkconfig` file is no longer created in my workspace. "git status" no longer works after I save a change, forcing me to manually copy settings from my build folder? **My opinion:** this pr should be reverted.

**Comment by brianignacio5** [comment]
This multi config is not the default use case of ESP-IDF. Not at all. In fact is a very specific use case. You don't need to use the project configuration editor at all. Previous behavior still works well.

I just trying to explain to you that `sdkconfig` is a file generated by the build system. See the esp-idf examples and you will notice that `sdkconfig` file doesn't exist. it is generated after running `idf.py menuconfig`.

This example (and the idea of project configuration in general) is to create multiple projects with a single source code. Let me give you an example use case. Say you have project using multiple tft screens. You want product 1 to use a screen A while product 2 to use a screen B. In your code you write a Kconfig file to choose between screen based on a sdkconfig value say CONFIG_SCREEN_TO_USE.

In the ESP-IDF documentation about [sdkconfig defaults](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/kconfig.html#using-sdkconfig-defaults)

Basically use sdkconfig.defaults as generators to have multiple sdkconfig files.

`[sdkconfig.common, sdkconfig.prod1]` -> `${workspaceFolder}/sdkconfig.product1`

`[sdkconfig.common, sdkconfig.prod2]` -> `${workspaceFolder}/sdkconfig.product2`

This feature is not intended at all to be used for most ESP-IDF projects.

**Comment by chipweinberger** [comment]
Thanks for your comments Brian. 

> This feature is not intended at all to be used for most ESP-IDF projects.

Having a "dev" and "release" mode is common use case IMO. Most products need these. This is all I want. 

> I just trying to explain to you that sdkconfig is a file generated by the build system. 

Yes I do understand that.

> Previous behavior still works well.

Before the `sdkconfig` file was created & edited in the workspace. Now it no longer does that anymore. Am I wrong?

**Comment by chipweinberger** [comment]
**Idea:** Perhaps you should create an example that has "dev" and "release" modes.

Show:

1. how to setup "dev" and "release" in the settings
2. how to edit "dev" using the "SDK Config Editor" & save the new dev configuration to your repo
3. how to edit "release" using the "SDK Config Editor" & save the new release configuration in your repo
4. how to build & flash "dev"
5. how to build & flash "release"

**Comment by brianignacio5** [comment]
> Before the sdkconfig file was created & edited in the workspace. Now it no longer does that anymore. Am I wrong?

The sdkconfig file is created and edited in the workspace by default in CMake projects and most ESP-IDF projects.

Here is extending the use of a single sdkconfig to use many sdkconfig files for a single code. You can switch between one and another switch configuration profile. For this specific esp-idf example, build system multiple configuration example, the sdkconfig default path is modified in [here](https://github.com/espressif/esp-idf/blob/master/examples/build_system/cmake/multi_config/CMakeLists.txt#L6)

**Comment by brianignacio5** [comment]
I just want to add that the original idea of the feature is for make easier scripting of these idf.py build and other commands.

While sdkconfig and sdkconfig defaults is one of many things we can do with `idf.py -B build_path -DSDKCONFIG=sdkconfig_path -DSDKCONFIG_DEFAULTS=[common,prod1]` we can do many other things based on the arguments here. I think we really need to clarify how all these translate to CLI commands.

**Comment by chipweinberger** [comment]
Any progress on https://github.com/espressif/vscode-esp-idf-extension/pull/870#issuecomment-1525182598?

**Comment by brianignacio5** [comment]
Sorry for late reply, I haven't had bandwidth to continue writing these docs but for sure I'll make time before next release and I'll tag you in the MR .

**Comment by chipweinberger** [comment]
I appreciate your communications! Thanks Brian.

**Comment by chipweinberger** [comment]
But want to bump this thread again. Looking forward to the docs for https://github.com/espressif/vscode-esp-idf-extension/pull/870#issuecomment-1525182598

**Comment by brianignacio5** [comment]
Could you check this PR with the docs ? https://github.com/espressif/vscode-esp-idf-extension/pull/983

The idf.py save-defconfig works on ESP-IDF v5.0

**Comment by chipweinberger** [comment]
overall looks good!

i left some comments and suggestions!
